File: javascript/src/Behavioral/Iterator/DvdList.js
(function() {
  define(['Behavioral/Iterator/DvdListIterator'], function(DvdListIterator) {
    var DvdList;
    return DvdList = (function() {
      function DvdList(titles, titleCount, arraySize) {
        this.titles = titles != null ? titles : [];
        this.titleCount = titleCount != null ? titleCount : 0;
        this.arraySize = arraySize != null ? arraySize : 0;
        this.iterator = null;
      }

      DvdList.prototype.append = function(title) {
        this.titles.push(title);
        return this.titleCount++;
      };

      DvdList.prototype.remove = function(title) {
        var founds, i, t, _i;
        founds = (function() {
          var _i, _len, _ref, _results;
          _ref = this.titles;
          _results = [];
          for (i = _i = 0, _len = _ref.length; _i < _len; i = ++_i) {
            t = _ref[i];
            if (title === t) {
              _results.push(i);
            }
          }
          return _results;
        }).call(this);
        if (founds.length > 0) {
          for (_i = founds.length - 1; _i >= 0; _i += -1) {
            i = founds[_i];
            this.titleCount--;
          }
          return this.titles.splice(i, 1);
        }
      };

      DvdList.prototype.createIterator = function() {
        if (this.iterator) {
          return this.iterator;
        }
        this.iterator = new DvdListIterator(this);
        return this.iterator;
      };

      return DvdList;

    })();
  });

}).call(this);

File: csharp/src/Behavioral/Iterator/DvdList.cs
//DvdList.java - the Concrete Aggregate (with a Concrete Iterator inner class)


namespace Behavioral.Iterator {


using System;
using System.Collections.Generic;


public class DvdList {  


   public List<string> titles;  
   //Yes, it would be easier to do this whole example with ArrayList
   // and ListIterator, but it certainly wouldn't be as much fun!   
   public int titleCount;
   //title count is always a real count of titles, but one ahead of
   //itself as a subscript
   public int arraySize;
    
   public DvdList() {
        titles = new List<string>();
        //using 3 to demonstrate array expansion more easily,
        //  not for efficency        
        titleCount = 0;
        arraySize = 3;
   }    
   
   public int count() {
       return titleCount;
   }
   
   public void append(string titleIn) {
		titles.Add(titleIn);
		titleCount++;
   }
   
   public void delete(string titleIn) {
       bool found = false;
       for (int i = 0; i < (titleCount -1); i++) {
           if (found == false) {
               if (titles[i].Equals(titleIn)) {
                   found = true;
                   titles[i] = titles[i + 1];
               }
           } else {
               if (i < (titleCount -1)) {
                   titles[i] = titles[i + 1];
               } else {
                   titles[i] = null;
               }
           }
       }
       
       if (found == true) {
           --titleCount;
       }
   }
   
   
   public DvdListIterator createIterator() {
       return new DvdListIteratorImpl(this);
   }
   
}

}
